### PR TITLE
minimum knapsack

### PR DESCRIPTION
**Pull request by Mathemalsky**
This is the first algorithm in where we could enter the case of an infeasible instance.
Do you think we should introduce own error types for that purpose inheriting from the errors from `std::except`?

### CODE REVIEW

**Comment by MilchRatchet** [comment]
> Do you think we should introduce own error types for that purpose inheriting from the errors from `std::except`?

I think we should not throw errors at all but rather return that the result is "infeasible". If I am not mistaken, an error ends the application right? If we simply return infeasibility then the user can decide on his own what to do with it. I had not thought about that in the previous algorithms but I think throwing errors in such a library is simply bad design.

File: src/AAL/AAL_knapsack.cpp [hunk 1]
```diff
@@ -118,3 +120,64 @@ AAL_knapsackResult AAL_knapsack_fptas(const AAL_knapsackInstance instance, const
...
 }
+
+AAL_knapsackResult AAL_knapsack_min(const AAL_knapsackInstance instance) {
+  // copy sizes and values into vectors
+  std::vector<unsigned int> zeros;
+  std::vector<float> sizes(instance.numberOfItems), values(instance.numberOfItems);
+  for (unsigned int i = 0; i < instance.numberOfItems; ++i) {
+    sizes[i]  = instance.itemSizes[i];
+    values[i] = instance.itemValues[i];
+    if (values[i] == 0) {
```
Comment (MilchRatchet) [inline src/AAL/AAL_knapsack.cpp]:
```suggestion
    if (values[i] == 0.0f) {
```

File: src/AAL/AAL_knapsack.cpp [hunk 2]
```diff
@@ -118,3 +120,64 @@ AAL_knapsackResult AAL_knapsack_fptas(const AAL_knapsackInstance instance, const
...
+
+  std::vector<unsigned int> A, A_complement(instance.numberOfItems);
+  std::iota(A_complement.begin(), A_complement.end(), 0);
+  // items of value 0 will be no longer considered
+  for (int i = zeros.size() - 1; i >= 0; ++i) {
+    std::swap(A_complement[A_complement.size() - 1], A_complement[zeros[i]]);
+    A_complement.pop_back();
+  }
+
+  float value = 0;
```
Comment (MilchRatchet) [inline src/AAL/AAL_knapsack.cpp]:
```suggestion
  float value = 0.0f;
```

File: src/AAL/AAL_knapsack.cpp [hunk 3]
```diff
@@ -118,3 +120,64 @@ AAL_knapsackResult AAL_knapsack_fptas(const AAL_knapsackInstance instance, const
...
+
+    // reduce the sizes by the value times the dual variable
+    for (unsigned int i = 0; i < A_complement.size(); ++i) {
+      sizes[A_complement[i]] -= min * values[A_complement[i]];
+    }
+  }
+
+  // construct the solution
+  unsigned int* items  = (unsigned int*) malloc(A.size() * sizeof(unsigned int));
+  float objectiveValue = 0;
```
Comment (MilchRatchet) [inline src/AAL/AAL_knapsack.cpp]:
```suggestion
  float objectiveValue = 0.0f;
```

**Review by MilchRatchet — CHANGES_REQUESTED** [review]
Some small immediate values being declared as an int when they should be a float. Besides that this is fine.

**Comment by Mathemalsky** [comment]
> > Do you think we should introduce own error types for that purpose inheriting from the errors from `std::except`?
> 
> I think we should not throw errors at all but rather return that the result is "infeasible". If I am not mistaken, an error ends the application right? If we simply return infeasibility then the user can decide on his own what to do with it. I had not thought about that in the previous algorithms but I think throwing errors in such a library is simply bad design.

No on that point you are mistaken. Errors are introduced exactly for that purpose: Reporting that something went wrong and you can decide how to deal with that information. Stopping the application is only the default if you don't use the error information.
This page gives a good overview of exceptions: https://docs.microsoft.com/en-us/cpp/cpp/errors-and-exception-handling-modern-cpp?view=msvc-160

But I'm not sure what happens if a `c` program receives an error from a `c++` library since as far as I know there is no error handling in `c`.

**Comment by MilchRatchet** [comment]
> 
> 
> > > Do you think we should introduce own error types for that purpose inheriting from the errors from `std::except`?
> > 
> > 
> > I think we should not throw errors at all but rather return that the result is "infeasible". If I am not mistaken, an error ends the application right? If we simply return infeasibility then the user can decide on his own what to do with it. I had not thought about that in the previous algorithms but I think throwing errors in such a library is simply bad design.
> 
> No on that point you are mistaken. Errors are introduced exactly for that purpose: Reporting that something went wrong and you can decide how to deal with that information. Stopping the application is only the default if you don't use the error information.
> This page gives a good overview of exceptions: https://docs.microsoft.com/en-us/cpp/cpp/errors-and-exception-handling-modern-cpp?view=msvc-160
> 
> But I'm not sure what happens if a `c` program receives an error from a `c++` library since as far as I know there is no error handling in `c`.

I see, so there is a try/except/finally kind of way of error handling. I had a look at it and it seems that the internal implementation of throwing and catching is based on the compiler and in fact, throwing an error is not even necessarily cross compatible over different compilers. Things apparently get even worse once you mix different operating systems or languages. Also you are right when you say that `C` and other languages have no error handling like that. Those languages would not be able to catch the errors. Thus I would guess not throwing an error and instead signaling this in the return value seems to be the correct way.

Afterall, if I look at different libraries I have used in the past (outside of `C#`) they have always returned some special value instead of throwing an error.

**Comment by Mathemalsky** [comment]
> > > > Do you think we should introduce own error types for that purpose inheriting from the errors from `std::except`?
> > > 
> > > 
> > > I think we should not throw errors at all but rather return that the result is "infeasible". If I am not mistaken, an error ends the application right? If we simply return infeasibility then the user can decide on his own what to do with it. I had not thought about that in the previous algorithms but I think throwing errors in such a library is simply bad design.
> > 
> > 
> > No on that point you are mistaken. Errors are introduced exactly for that purpose: Reporting that something went wrong and you can decide how to deal with that information. Stopping the application is only the default if you don't use the error information.
> > This page gives a good overview of exceptions: https://docs.microsoft.com/en-us/cpp/cpp/errors-and-exception-handling-modern-cpp?view=msvc-160
> > But I'm not sure what happens if a `c` program receives an error from a `c++` library since as far as I know there is no error handling in `c`.
> 
> I see, so there is a try/except/finally kind of way of error handling. I had a look at it and it seems that the internal implementation of throwing and catching is based on the compiler and in fact, throwing an error is not even necessarily cross compatible over different compilers. Things apparently get even worse once you mix different operating systems or languages. Also you are right when you say that `C` and other languages have no error handling like that. Those languages would not be able to catch the errors. Thus I would guess not throwing an error and instead signaling this in the return value seems to be the correct way.
> 
> Afterall, if I look at different libraries I have used in the past (outside of `C#`) they have always returned some special value instead of throwing an error.

So our way to go is including an status `int` in the structs we return?

**Comment by MilchRatchet** [comment]
> 
> 
> > > > > Do you think we should introduce own error types for that purpose inheriting from the errors from `std::except`?
> > > > 
> > > > 
> > > > I think we should not throw errors at all but rather return that the result is "infeasible". If I am not mistaken, an error ends the application right? If we simply return infeasibility then the user can decide on his own what to do with it. I had not thought about that in the previous algorithms but I think throwing errors in such a library is simply bad design.
> > > 
> > > 
> > > No on that point you are mistaken. Errors are introduced exactly for that purpose: Reporting that something went wrong and you can decide how to deal with that information. Stopping the application is only the default if you don't use the error information.
> > > This page gives a good overview of exceptions: https://docs.microsoft.com/en-us/cpp/cpp/errors-and-exception-handling-modern-cpp?view=msvc-160
> > > But I'm not sure what happens if a `c` program receives an error from a `c++` library since as far as I know there is no error handling in `c`.
> > 
> > 
> > I see, so there is a try/except/finally kind of way of error handling. I had a look at it and it seems that the internal implementation of throwing and catching is based on the compiler and in fact, throwing an error is not even necessarily cross compatible over different compilers. Things apparently get even worse once you mix different operating systems or languages. Also you are right when you say that `C` and other languages have no error handling like that. Those languages would not be able to catch the errors. Thus I would guess not throwing an error and instead signaling this in the return value seems to be the correct way.
> > Afterall, if I look at different libraries I have used in the past (outside of `C#`) they have always returned some special value instead of throwing an error.
> 
> So our way to go is including an status `int` in the structs we return?

Yes.

**Pull request merged**